### PR TITLE
#156 implement rest env api in ruby v3

### PR DESCRIPTION
**Pull request by cdeler**
@lopopolo Yet another piece of ENV object has been done

### CODE REVIEW

File: artichoke-backend/src/extn/core/env/env.rb [hunk 1]
```diff
@@ -36,10 +36,58 @@ def delete_if
...
+  def fetch(name, default = nil)
+    value = self[name]
+
+    return value unless value.nil?
+
+    return yield name if block_given?
+
+    return default unless default.nil?
+
+    raise KeyError, 'Variable ' + name + " doesn't exist"
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: can we convert this to use `String` interpolation?

It also appears MRI uses a different error message: `"key not found: #{name}"`

```
[2.6.3] > ENV.fetch('aaaaaaaa')
Traceback (most recent call last):
        6: from /usr/local/var/rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        5: from /usr/local/var/rbenv/versions/2.6.3/bin/irb:23:in `load'
        4: from /usr/local/var/rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):11
        2: from (irb):11:in `rescue in irb_binding'
        1: from (irb):11:in `fetch'
KeyError (key not found: "aaaaaaaa")
```

Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: Sure, I've fixed a remark

File: artichoke-backend/src/extn/core/env/env.rb [hunk 2]
```diff
@@ -36,10 +36,58 @@ def delete_if
...
   end
 
+  def fetch(name, default = nil)
+    value = self[name]
+
+    return value unless value.nil?
+
+    return yield name if block_given?
+
+    return default unless default.nil?
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: ENV.fetch with an explicit nil default returns nil for a not found key.

One way we can implement this is with an expression as the default value:

```ruby
def fetch(name, default = (not_set = true))
  # ...
end
```

If no default is supplied, the expression is evaluated. Then we can `return default if not_set.nil?`

Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: Marvelous. Ruby is so tricky :-)

Fixed and compared with the default `irb` implementation.

**Pull request merged**